bftengine/src/bftengine/SeqNumInfo.cpp

```diff
@@ -266,7 +289,135 @@ Time SeqNumInfo::getTimeOfFirstRelevantInfoFromPrimary() const { return firstSee
+      result = fastPathThresholdCollector->addMsgWithPartialSignature(m, myId);
+      break;
+    default:
+      ConcordAssert(false && "Unhandled CommitPath!");
+  }
+
+  return result;
+}
+
+bool SeqNumInfo::addFastPathPartialCommitMsg(PartialCommitProofMsg* m) {
+  ConcordAssert(m != nullptr);
+
+  const ReplicaId repId = m->senderId();
+  ConcordAssert(repId != replica->getReplicasInfo().myId());
+  ConcordAssert(replica->getReplicasInfo().isIdOfReplica(repId));
+  // PartialCommitProofMsg is allowed before a prePrepare is received
+  ConcordAssert(!prePrepareMsg || hasMatchingPrePrepare(m->seqNumber()));
+
+  if (hasFastPathFullCommitProof()) return false;
+
+  const auto* selfPartialCommitProof = getFastPathSelfPartialCommitProofMsg();
+  const CommitPath cPath = m->commitPath();
+
+  // Check different fast commit path
+  if (selfPartialCommitProof && selfPartialCommitProof->commitPath() != cPath) {
+    LOG_WARN(CNSUS,
+             "Ignoring PartialCommitProofMsg (" << CommitPathToStr(cPath) << "). Current path is "
+                                                << CommitPathToStr(selfPartialCommitProof->commitPath()));
+    return false;
+  }
+
+  auto result = false;
+  switch (cPath) {
+    case CommitPath::OPTIMISTIC_FAST:
+      result = fastPathOptimisticCollector->addMsgWithPartialSignature(m, repId);
+      break;
+    case CommitPath::FAST_WITH_THRESHOLD:
+      result = fastPathThresholdCollector->addMsgWithPartialSignature(m, repId);
+      break;
+    default:
+      ConcordAssert(false && "Unhandled CommitPath!");
+  }
+
+  return result;
+}
+
+bool SeqNumInfo::addFastPathFullCommitMsg(FullCommitProofMsg* m, bool directAdd) {
+  ConcordAssert(m != nullptr);
+
+  if (hasFastPathFullCommitProof()) return false;
+
+  PartialCommitProofMsg* myPCP = getFastPathSelfPartialCommitProofMsg();
+
+  if (myPCP == nullptr) {
+    // TODO(GG): can be improved (we can keep the FullCommitProof  message until myPCP!=nullptr
+    LOG_WARN(CNSUS, "FullCommitProofMsg arrived before PrePrepare. TODO(GG): should be handled to avoid delays. ");
+    return false;
+  }
+
+  if (m->seqNumber() != myPCP->seqNumber() || m->viewNumber() != myPCP->viewNumber()) {
+    LOG_WARN(CNSUS, "Received unexpected FullCommitProofMsg");
+    return false;
+  }
+
+  // ToDo--EDJ: FullCommitProofMsg doesn't say to which fast path it belongs
+  const auto cPath = myPCP->commitPath();
+  bool result = false;
+  switch (cPath) {
+    case CommitPath::OPTIMISTIC_FAST:
+      result = directAdd ? fastPathOptimisticCollector->initMsgWithCombinedSignature(m)
+                         : fastPathOptimisticCollector->addMsgWithCombinedSignature(m);
+      break;
+    case CommitPath::FAST_WITH_THRESHOLD:
+      result = directAdd ? fastPathThresholdCollector->initMsgWithCombinedSignature(m)
+                         : fastPathThresholdCollector->addMsgWithCombinedSignature(m);
+      break;
+    default:
+      ConcordAssert(false && "Unhandled CommitPath!");
```

Comment:
Better to use LOG_ERROR and ConcordAssert(false)

Reply:
I've split the ConcordAsserts as suggested since this is the more common usage. Also the log now includes the invalid commit path value that caused the error.